perllib/FixMyStreet/App/Controller/Report/New.pm

```diff
@@ -1543,55 +1544,63 @@ sub process_confirmation : Private {
+        my @problems = ($problem);
+        if (my $grouped_ids = $problem->get_extra_metadata('grouped_ids')) {
+            foreach my $id (@$grouped_ids) {
+                my $problem = $c->model('DB::Problem')->find({ id => $id }) or next;
+                push @problems, $problem;
+            }
         }
-    }
-    foreach my $problem (@problems) {
-        $problem->confirm;
-        $problem->update({ lastupdate => \'current_timestamp' });
-
-        # Subscribe problem reporter to email updates
-        $c->forward( '/report/new/create_related_things', [ $problem ] );
-    }
+        foreach my $problem (@problems) {
+            $problem->confirm;
+            $problem->update({ lastupdate => \'current_timestamp' });
 
-    # log the problem creation user in to the site
-    if ( $data->{name} || $data->{password} ) {
-        if (!$problem->user->email_verified) {
-            $problem->user->email( $data->{email} ) if $data->{email};
-        } elsif (!$problem->user->phone_verified) {
-            $problem->user->phone( $data->{phone} ) if $data->{phone};
-        }
-        $problem->user->password( $data->{password}, 1 ) if $data->{password};
-        for (qw(name title facebook_id twitter_id)) {
-            $problem->user->$_( $data->{$_} ) if $data->{$_};
+            # Subscribe problem reporter to email updates
+            $c->forward( '/report/new/create_related_things', [ $problem ] );
         }
-        $problem->user->add_oidc_id($data->{oidc_id}) if $data->{oidc_id};
-        $problem->user->extra({
-            %{ $problem->user->get_extra() },
-            %{ $data->{extra} }
-        }) if $data->{extra};
-
-        $problem->user->update;
-
-        # Make sure extra oauth state is restored, if applicable
-        foreach (qw/logout_redirect_uri change_password_uri/) {
-            if ($data->{$_}) {
-                $c->session->{oauth} ||= ();
-                $c->session->{oauth}{$_} = $data->{$_};
+        $data->{redeemed} = time();
+        $token->update( {data => $data} );
+
+        # update user details
+        if ( $data->{name} || $data->{password} ) {
+            if (!$problem->user->email_verified) {
+                $problem->user->email( $data->{email} ) if $data->{email};
+            } elsif (!$problem->user->phone_verified) {
+                $problem->user->phone( $data->{phone} ) if $data->{phone};
             }
+            $problem->user->password( $data->{password}, 1 ) if $data->{password};
+            for (qw(name title facebook_id twitter_id)) {
+                $problem->user->$_( $data->{$_} ) if $data->{$_};
+            }
+            $problem->user->add_oidc_id($data->{oidc_id}) if $data->{oidc_id};
+            $problem->user->extra({
+                %{ $problem->user->get_extra() },
+                %{ $data->{extra} }
+            }) if $data->{extra};
+            $problem->user->update;
         }
     }
+
+    # Make sure extra oauth state is restored, if applicable
```

Comment:
I think this can be done only the once? It updates the session server-side, unless I've missed something. (Also assume there might have been a reason to only do this if name/password set :) )

Reply:
This is sill present, this block has moved outside the `if ( $data->{name} || $data->{password} ) {` block which seems unrelated to the changes.